ea-2022-04-ndvi-automation.py

```diff
@@ -0,0 +1,645 @@
+# Function A
+def open_clean_bands(band_path,
+                     valid_range=None):
+    """Open and mask a single Landsat band using a pixel_qa layer.
+
+    Parameters
+    ----------
+    band_path : string
+        A path to the array to be opened
+    valid_range : tuple (optional)
+        A tuple of min and max range values for the data. Default = None
+
+    Returns
+    ----------
+    arr : xarray DataArray
+        An xarray DataArray with values that should be masked set to 1 for
+        True (Boolean)
+    """
+    # Open, clip, mask single band using rioxarray to valid range
+    band = (rxr.open_rasterio(band_path, masked=True)
+            .rio.clip(crop_bound.geometry, from_disk=True)
+            .squeeze())
+    
+    # Specify the valid range of values
+    if valid_range: 
+        mask = ((band <= 0) | (band > 10000))
+        band = band.where(~mask, np.nan)
+        
+    return band
+
+# Function B
+def mask_crop_ndvi(all_bands,
+                   crop_bound,
+                   pixel_qa,
+                   vals):
+    """Open and mask a single Landsat band using a pixel_qa layer.
+
+    Parameters
+    -----------
+    all_bands : list
+        A list containing paths to Landsat bands 4 and 5 as .tif files
+    pixel_qa : xarray DataArray
+        An xarray DataArray with pixel qa values that have not yet been turned into a mask (0s and 1s)
+    crop_bounds : geopandas GeoDataFrame
+        A geopandas dataframe to be used to crop the raster data using
+        rasterio mask()
+    vals : list
+        A list of values needed to create the cloud mask
+
+    Returns
+    --------
+    mean_ndvi : xarray.DataArray
+        A cropped, masked xarray object containing NDVI values
+
+    """
+    # Create empty list
+    bands = []
+    for band_path in all_bands_path:
+        clean_bands = open_clean_bands(band_path=band_path,
+                                valid_range=(0, 10000))
+
+        bands.append(clean_bands)
+
+    # Open and clip cloud mask layer
+    cl_mask = (rxr.open_rasterio(pixel_qa_path, masked=True)
+               .rio.clip(crop_bound.geometry, from_disk=True).squeeze())
+
+    # Calculate NDVI
+    ndvi_xr = (bands[1]-bands[0]) / (bands[1]+bands[0])
+
+    # Apply cloud mask to NDVI
+    all_masked_values = [328, 392, 840, 904, 1350, 352, 368, 416,
+                         432, 480, 864, 880, 928, 944, 992, 480, 992]
+
+    ndvi_mask = ndvi_xr.where(~cl_mask.isin(all_masked_values))
+
+    # Calculate NDVI values
+    mean_ndvi = ndvi_mask.mean(skipna=True).item()
+
+    return mean_ndvi
+
+
+# In[8]:
+
+
+# Define directory name
+site_crop_dir = "landsat-crop"
+
+# Create empty list
+ndvi_list = []
+
+# Loop through each site directory
+for site_dir in all_sites_dirs:
+    print("I am looping through: ", site_dir)
+    # Get site name
+    site = os.path.normpath(site_dir).split(os.sep)[-1]
+
+    # Get a list of subdirectories for the site
+    print("I am working on", site, "field site now")
```

Comment:
I know we've only used print statements, so this is fine! If you're looking for something better, I'd encourage you to look into the logging package because as you start using more data it's a much better tool for sending messages to exactly where you want them.

Reply:
Interesting, conceptually I think it would be useful to have descriptive strings within the processed outputs so one can see the code does.